### PR TITLE
diff image config

### PR DESCRIPTION
**Pull request by XavierCheng1996**
compare the image config now with just result of whether configs are same or not.

### CODE REVIEW

File: bin/makisu/cmd/diff.go [hunk 1]
```diff
@@ -95,10 +96,23 @@ func (cmd *diffCmd) Diff(imagesFullName []string) error {
...
+		// Check image config.
+		imageConfig, err := client.PullImageConfig(manifest.GetConfigDigest())
+		if err != nil {
+			panic(err)
+		}
+		imageConfigs = append(imageConfigs, imageConfig)
+	}
+
+	log.Infof("* Diff two images config")
+	if os.SameFile(imageConfigs[0], imageConfigs[1]) {
```
Comment (yiranwang52) [inline bin/makisu/cmd/diff.go:109]: This always false.

File: bin/makisu/cmd/diff.go [hunk 2]
```diff
@@ -95,10 +96,23 @@ func (cmd *diffCmd) Diff(imagesFullName []string) error {
...
+		log.Infof("Image config files are same")
+	} else {
+		log.Infof("Image config files are different")
```
Comment (yiranwang52) [inline bin/makisu/cmd/diff.go:112]: Print content in this case.

Comment (XavierCheng1996) [reply bin/makisu/cmd/diff.go:112]: ok.

Comment (XavierCheng1996) [reply bin/makisu/cmd/diff.go:112]: after second look. As for the config, we would just have `fileInfo` which doesn't tell much. we don't have sth like `memFS` to untar the corresponding tar balls to a file system. Should we just tells some basic information of config like `size, name and etc`.

File: bin/makisu/cmd/diff.go [hunk 3]
```diff
@@ -94,11 +99,36 @@ func (cmd *diffCmd) Diff(imagesFullName []string) error {
...
+		}
+
+		config := new(image.Config)
+		if err := json.Unmarshal(configBytes, config); err != nil {
+			panic(fmt.Errorf("unmarshal image%d config file %s: %s", i+1, manifest.GetConfigDigest().Hex(), err))
+		}
+		imageConfigs = append(imageConfigs, config)
+	}
+
+	log.Infof("* compare image %s and image %s config", pullImages[0].GetRepository()+":"+pullImages[0].GetTag(), pullImages[1].GetRepository()+":"+pullImages[1].GetTag())
```
Comment (yiranwang52) [inline bin/makisu/cmd/diff.go:121]:
```
* Diff image configs *
```

File: bin/makisu/cmd/diff.go [hunk 4]
```diff
@@ -94,11 +99,36 @@ func (cmd *diffCmd) Diff(imagesFullName []string) error {
...
+	if configDiff := cmp.Diff(imageConfigs[0], imageConfigs[1], cmpopts.IgnoreUnexported(image.Config{})); configDiff != "" {
+		// Format the diff string.
+		configDiff = strings.ReplaceAll(configDiff, "\n", "")
+		configDiff = strings.ReplaceAll(configDiff, "\t", "")
+		configDiff = strings.ReplaceAll(configDiff, "\"", "")
+		configDiff = strings.ReplaceAll(configDiff, "  ", "")
+		log.Infof("-image %s +image %s):\n%s", pullImages[0].GetRepository()+":"+pullImages[0].GetTag(), pullImages[1].GetRepository()+":"+pullImages[1].GetTag(), configDiff)
 	}
 
 	log.Infof("* Diff two images")
```
Comment (yiranwang52) [inline bin/makisu/cmd/diff.go:131]:
```
* Diff image layers*
```

File: bin/makisu/cmd/diff.go [hunk 5]
```diff
@@ -94,11 +99,36 @@ func (cmd *diffCmd) Diff(imagesFullName []string) error {
...
+		if err := json.Unmarshal(configBytes, config); err != nil {
+			panic(fmt.Errorf("unmarshal image%d config file %s: %s", i+1, manifest.GetConfigDigest().Hex(), err))
+		}
+		imageConfigs = append(imageConfigs, config)
+	}
+
+	log.Infof("* compare image %s and image %s config", pullImages[0].GetRepository()+":"+pullImages[0].GetTag(), pullImages[1].GetRepository()+":"+pullImages[1].GetTag())
+	if configDiff := cmp.Diff(imageConfigs[0], imageConfigs[1], cmpopts.IgnoreUnexported(image.Config{})); configDiff != "" {
+		// Format the diff string.
+		configDiff = strings.ReplaceAll(configDiff, "\n", "")
```
Comment (yiranwang52) [inline bin/makisu/cmd/diff.go:124]: Maybe we should stop using json logger? then there might not be need for these replacements.

Comment (XavierCheng1996) [reply bin/makisu/cmd/diff.go:124]: the configDiff itself has many weird `\n \t` characters... so I think it kind of necessary to format it.

Comment (yiranwang52) [reply bin/makisu/cmd/diff.go:124]: If we don't print if as part of a json log line it should look better.

**Pull request merged**